src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -1593,6 +1625,47 @@ public void PublicApi_InOutAttributeMissing()
         }
 
+        [Theory]
+        [InlineData("[Cdecl]", SignatureCallingConvention.CDecl)]
+        [InlineData("[Thiscall]", SignatureCallingConvention.ThisCall)]
+        [InlineData("[Stdcall]", SignatureCallingConvention.StdCall)]
+        [InlineData("[Fastcall]", SignatureCallingConvention.FastCall)]
+        [InlineData("", SignatureCallingConvention.Unmanaged)]
+        public void PublicApi_CallingConventions_NoModopts(string expectedText, SignatureCallingConvention convention)
+        {
+            var comp = (Compilation)CreateCompilation("");
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(@string, returnRefKind: RefKind.None, parameterTypes: ImmutableArray<ITypeSymbol>.Empty, parameterRefKinds: ImmutableArray<RefKind>.Empty, convention);
+            AssertEx.Equal($"delegate* unmanaged{expectedText}<System.String>", ptr.ToTestDisplayString());
+            ptr = comp.CreateFunctionPointerTypeSymbol(@string, returnRefKind: RefKind.RefReadOnly, parameterTypes: ImmutableArray<ITypeSymbol>.Empty, parameterRefKinds: ImmutableArray<RefKind>.Empty, convention);
+            AssertEx.Equal($"delegate* unmanaged{expectedText}<ref readonly modreq(System.Runtime.InteropServices.InAttribute) System.String>", ptr.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_CallingConventions_Modopts()
+        {
+            var comp = (Compilation)CreateCompilation("");
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+            var cdeclType = comp.GetTypeByMetadataName("System.Runtime.CompilerServices.CallConvCdecl");
+            var stdcallType = comp.GetTypeByMetadataName("System.Runtime.CompilerServices.CallConvStdcall");
+            Assert.NotNull(cdeclType);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(@string, returnRefKind: RefKind.None, parameterTypes: ImmutableArray<ITypeSymbol>.Empty, parameterRefKinds: ImmutableArray<RefKind>.Empty, SignatureCallingConvention.Unmanaged, ImmutableArray.Create(cdeclType, stdcallType)!);
+            AssertEx.Equal("delegate* unmanaged[Cdecl, Stdcall]<System.String modopt(System.Runtime.CompilerServices.CallConvCdecl) modopt(System.Runtime.CompilerServices.CallConvStdcall)>", ptr.ToTestDisplayString());
+            ptr = comp.CreateFunctionPointerTypeSymbol(@string, returnRefKind: RefKind.RefReadOnly, parameterTypes: ImmutableArray<ITypeSymbol>.Empty, parameterRefKinds: ImmutableArray<RefKind>.Empty, SignatureCallingConvention.Unmanaged, ImmutableArray.Create(cdeclType, stdcallType)!);
+            AssertEx.Equal("delegate* unmanaged[Cdecl, Stdcall]<ref readonly modopt(System.Runtime.CompilerServices.CallConvCdecl) modopt(System.Runtime.CompilerServices.CallConvStdcall) modreq(System.Runtime.InteropServices.InAttribute) System.String>", ptr.ToTestDisplayString());
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider testing `CreateFunctionPointerTypeSymbol(...)` with just one of the two modopt types. Presumably the result should still include `modopt(...)` even though the (equivalent?) function pointer could be created by passing a specific `SignatureCallingConvention` instead.

Reply:
They're not equivalent: the SignatureCallKind will be `Unmanaged`, so they don't be considered equivalent and there's no way to write such a function pointer in source. It's a good test though

---
In reply to: [467367613](https://github.com/dotnet/roslyn/pull/46467#discussion_r467367613) [](ancestors = 467367613)